modules/basics/src/main/java/com/opengamma/basics/currency/FxMatrix.java

```diff
@@ -0,0 +1,557 @@
+      // unused rows
+      return new FxMatrix(ImmutableMap.copyOf(currencies), copyArray(rates, currencies.size()));
+    }
+
+    /**
+     * Adds a new rate for a currency pair to the builder. See
+     * {@link #addRate(Currency, Currency, double)} for full
+     * explanation.
+     *
+     * @param currencyPair  the currency pair to be added
+     * @param rate  the FX rate between the base currency of the pair and the
+     *   counter currency. The rate indicates the value of one unit of the base
+     *   currency in terms of the counter currency.
+     * @return the builder updated with the new rate
+     */
+    public Builder addRate(CurrencyPair currencyPair, double rate) {
+      ArgChecker.notNull(currencyPair, "currencyPair");
+      return addRate(currencyPair.getBase(), currencyPair.getCounter(), rate);
+    }
+
+    /**
+     * Add a new pair of currencies to the builder. There are a number
+     * of possible outcomes:
+     * <ul>
+     *   <li>
+     *     The builder is currently empty. In this case these currencies
+     *     and rates will be added as the initial pair.</li>
+     *   <li>
+     *     The builder is non-empty and neither of the currencies are
+     *     currently in the matrix. In this case the currencies cannot be
+     *     immediately added as there is no common currency to allow the
+     *     cross rates to be calculated. The currencies and rates are put
+     *     into a pending set for later processing, for example after another
+     *     pair containing one of the currencies and a currency already in
+     *     the matrix is added. If no such event occurs, then an exception
+     *     will be thrown whe {@link #build()} is called.</li>
+     *   <li>
+     *     The builder is non-empty and one of the currencies in the pair
+     *     is already in the matrix, whilst the other is not. In this case
+     *     the pair and rate is added to the matrix and all the cross rates
+     *     to the other currencies are calculated.
+     *   </li>
+     *   <li>
+     *     The builder is non-empty and contains both of the currencies in
+     *     the pair. In this case the pair is treated as an update to the
+     *     rate already in the matrix. The first currency (ccy1) is treated
+     *     as the reference currency and the second currency (ccy2) is the
+     *     updated currency. All rates involving the updated currency will
+     *     be recalculated using the new rate.
+     *     <p>
+     *     Note that due one of the rates being treated as a reference, this
+     *     operation is not symmetric. That is, the result of
+     *     {@code matrix.addRate(USD, EUR, 1.23)} will be different from the
+     *     result of {@code matrix.addRate(EUR, USD, 1 / 1.23)} when there
+     *     are other currencies present in the builder.
+     *   </li>
+     * </ul>
+     * An invocation of the method with {@code builder.addRate(GBP, USD, 1.6)}
```

Comment:
I think this paragraph is the most important part of the comment and should probably go at the top.